lib/evaluate/expression.h

```diff
@@ -811,18 +811,32 @@ template<> class Expr<SomeType> : public ExpressionBase<SomeType> {
 };
 
 // An assignment is either intrinsic (with lhs and rhs) or user-defined,
-// represented as a ProcedureRef.
+// represented as a ProcedureRef. A pointer assignment optionally also has
+// a bounds-spec or bounds-remapping.
 class Assignment {
 public:
-  explicit Assignment(ProcedureRef &&x) : u{std::move(x)} {}
-  Assignment(Expr<SomeType> &&lhs, Expr<SomeType> &&rhs)
-    : u{IntrinsicAssignment{std::move(lhs), std::move(rhs)}} {}
+  UNION_CONSTRUCTORS(Assignment)
   struct IntrinsicAssignment {
     Expr<SomeType> lhs;
     Expr<SomeType> rhs;
   };
-
-  std::variant<IntrinsicAssignment, ProcedureRef> u;
+  struct PointerAssignment {
+    // all upper bounds are present or all are absent
```

Comment:
The representation doesn't enforce this constraint, but it could, by splitting the data member `bounds` into `lowerBounds` and an optional `upperBounds`.

Reply:
That wouldn't enforce the constraint that if upperBounds is present it must have the same size as lower bounds.

Originally I had a variant of: nothing, a list of lower bounds, or a list of pairs. This enforces the constraints but made using it seem too complicated.

Reply:
How about a similar variant but without the empty case, since an empty vector of lower bounds could serve it?

Reply:
OK, my latest commit implements it that way. Let me know what you think.

Reply:
I like it, and it should be a good build-time defense against future problems.